Outline Rust project 4

brson opened the pull request:
Fixes https://github.com/pingcap/talent-plan/issues/38

I put this together quickly after our discussion @sticnarf. I do have a few concerns:

- there is apparently no production-quality concurrent map type we can incorporate :-/
- like in the previous project, this abstracts a thing then compares to implementations. I don't know if it is valuable to do the same type of exercise again. I'd prefer if we can find a new angle on this, so I've suggested we do a compile-time abstraction using cargo features, but I'm open to any ideas.

sticnarf commented:
@brson For the first concern, I search a second time and think it possible we use a concurrent hashmap.

A concurrent hashmap implementation that should be reliable is [chashmap](https://crates.io/crates/chashmap), but it is not actively maintained in the recent two years.

Another one I missed before because its name does not contain a keyword concurrent is [evmap](https://github.com/jonhoo/rust-evmap). It seems it is in active development.

Would you please evaluate them and possibly write them in a candidate list, then students can choose one from them.

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,92 @@
+- like with the sled comparison, abstract the thread
+  pool, and add an existing one
+- can we do a different _kind_ of abstraction than we did for sled? maybe lets
+  try compile time abstraction and cargo features
```
Yes, we can have a try. But I don't know what it'll exactly look like yet. Maybe we can add a "xxx-threadpool" features to the project and if we set the feature on through cargo command line arguments, we use the third-party threadpool instead of ours. However, I don't know if it is good or usual to switch implementations through feature flags.

brson replied:
Yeah, it feel unnatural to me to use compile time for this. I'm disinclined to do it. I mean, probably one wouldn't make the threadpool pluggable at all.

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,92 @@
+
+- read-centric workload, measuring network throughput
+- is just writing more criterion benchmarks useful?
+- can we come up with other tools to measure with,
```
The students use the client he wrote to benchmark the server. It can be a problem if the bottleneck is the client, for example, students use only one block client to test the server. So it is good if we can use a mature pressure testing tool to test. However, we use a custom protocol but most tools aim at HTTP. Thus, I think we should at least hint students that they should pay attention to concurrency at the client side if they are doing benchmarks.

sticnarf replied:
I have no idea about what else we can benchmark on about a thread pool now

brson replied:
Yes this is a problem I've encountered writing the text. The overhead of using the client is probably going to cause huge problems for benchmarking.

One thing I think we can benchmark is to once again test sled in a multi-threaded setting. It should perform markedly better than our engine.

brson commented:
@sticnarf I've added a bunch of text here, but still not finished. You should be able to get started.

mapleFU commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,447 @@
+
+At this point, if you've carried over your code from previous projects, your
+single-threaded `KvsServer` should work once again, but now with a `KvsEngine`
+that can later be shared across thrteads.
```
that can later be shared across thrteads.
                                                      ^^^^^^^^-> Maybe threads?

brson replied:
👍

mapleFU commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,447 @@
+revisit that 
+
+
+## Part 3: Abstract thread pools
```
Maybe it's part 5.

brson replied:
:+1

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,447 @@
+The library interface is nearly the same except for two things. First this time
+all the `KvsEngine`, `KvStore`, etc. methods take `&self` instead of `&mut
+self`, and now it implements `Clone`. This is common with parallel
+datastructions. Why is that? It's not that we're not going to be writing
```
```suggestion
data structures. Why is that? It's not that we're not going to be writing
```

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,447 @@
+
+## Project setup
+
+Create a new cargo project and copy `tests/tests.rs` into it. This project
```
```suggestion
Create a new cargo project and copy the `tests` directory into it. This project
```
We may have multiple test files now...

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,447 @@
+
+This gives us a lot of clues about the implementation strategy we're pursuing.
+First, think about why the engine needs to implement `Clone` when we have a
+multi-threaded implementation. Instead of being coy, this once I'll just tell
```
I believe "coy" is a new word for most Chinese people... Use "shy" instead?

brson replied:
Great suggestion.

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,447 @@
+
+[heap]: https://stackoverflow.com/questions/79923/what-and-where-are-the-stack-and-heap
+
+So, _hoist the date inside your implementation of `KvEngine`, `KvsStore` onto
```
I don't understand "hoist the date"... Is it a slang?

brson replied:
it should have said `data`, but "hoist" is slang. It means "pull up" or "pull out": "pull up your data from the stack to the heap". I'll change it.

brson replied:
more precisely, "hoist" isn't _slang_ but using it in this way is jargony.

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,447 @@
+[heap]: https://stackoverflow.com/questions/79923/what-and-where-are-the-stack-and-heap
+
+So, _hoist the date inside your implementation of `KvEngine`, `KvsStore` onto
+the heap usinga thread-safe shared pointer type and protect it behind a lock of
```
```suggestion
the heap using a thread-safe shared pointer type and protect it behind a lock of
```

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,447 @@
+syscalls is a common source of easy optimization. That stack then has to be
+carefully initialized so that first [stack _frame_] contains the appropriate
+values for the return pointer and other things. In Rust the stack needs to be
+configured with a [guard page]. That's two more syscalls, [to `mmap` and to
```
Seems that these two syscalls are saved in Linux. Add "except in Linux" for strictness?

brson replied:
:eyeroll: yeah, I saw that after I wrote this and read the std src. It's a bummer because I think this is an illustrative point.

It's also unfortunate that all the dirty mechanics of launching the thread are hidden in libpthread.

I'll do as you suggest.

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,447 @@
+
+Then at some point the OS performs a [context switch] onto the new thread. With
+a thread pool, all that setup overhead is only done for a few threads, and
+subsequent jobs are simply context switches into that pool of threads.
```
I think the cost of context switch and scheduling among a large number of threads should be mentioned too.

brson replied:
Can you say more about what you are thinking? I can imagine explaining that scheduling strategies themselves have a cost that has to be taken into consideration.

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,447 @@
+
+_Create the `RoundRobinThreadPool` type, implementing `ThreadPool`_.
+
+The tools at your disposal are `[thread::spawn]`, [`catch_unwind`], [`mpsc`]
```
It's OK to use `std::mpsc` to implement a round-robin thread pool. But I just realize that it is extremely easy to use mpmc crossbeam channel to implement a thread pool. Every thread just loops and receives tasks from the channel...

brson replied:
That would though not be round-robin, and I'm hoping that round-robin performs particularly badly in comparison to a production threadpool.

I would like to introduce crossbeam though, and it is much easier to write a threadpool with a single queue. Hm.

I'll change it and hope that we can find maybe a work-stealing threadpool that performs better. I have only heard bad things about every rust thread pool... but I'm planning to use rayon and hope for the best.

brson replied:
Maybe I'll try to leave the strategy open, have some background reading about scheduling strategies, and point to a few tools.

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,447 @@
+Again your `kvs-server` should still work the same as previously, now with a
+slightly more clever multi-threading model. This time you'll want to call
+the thread pool constructor with an appropriate number of threads. For
+now you can create a thread per CPU, using the [`num_threads`] crate. We'll
```
```suggestion
now you can create a thread per CPU, using the [`num_cpus`] crate. We'll
```

brson replied:
doh

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,447 @@
+different implementations, now you are going to use the `ThreadPool` abstraction
+to do the same.
+
+If you haven't already, add a second type parameter to `KvServer` to represent
```
```suggestion
If you haven't already, add a second type parameter to `KvsServer` to represent
```

brson commented:
Thanks for the review and all the typo catches. Someday I will learn to use a spellcheck.

brson commented:
I fixed review comments and rewrote the thread pool section to create a "SharedQueueThreadPool" instead. I tried to write it in a way that heavily hinted to investigate crossbeam without saying the good solution outright.

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,812 @@
+
+[heap]: https://stackoverflow.com/questions/79923/what-and-where-are-the-stack-and-heap
+
+So, _hoist the data inside your implementation of `KvEngine`, `KvsStore` onto
```
```suggestion
So, _hoist the data inside your implementation of `KvsEngine`, `KvStore` onto
```

sticnarf commented on rust/projects/parallelism/project.md:
```diff
@@ -0,0 +1,812 @@
+re-implement it as an extension to this project).
+
+Recall from the project spec that, this time, our `KvsEngine` takes `self` as
+`&self` instead of `&mut self` as previously, and it also implements `Clone`.
```
Should also implement `Send`

sticnarf commented on rust/projects/parallelism/project.md:
```suggestion
use a single shared queue to distribute work to idle threads. That means that
```

sticnarf commented on rust/projects/parallelism/project.md:
```suggestion
Messages in Rust are typically represented as enums, with variants for each
```

sticnarf commented on rust/projects/parallelism/project.md:
```suggestion
_Historical note: the existence of channels in Rust's standard library is a bit
```

brson commented:
All comments resolved. I've merged this in order to move on to other things and ahead of merging your code pr.
